subsys/bluetooth/host/l2cap.c

```diff
@@ -1295,15 +1293,20 @@ static int l2cap_chan_le_send(struct bt_l2cap_le_chan *ch, struct net_buf *buf,
 				       l2cap_chan_seg_sent, &ch->chan);
 	}
 
-	if (ret) {
-		BT_WARN("Unable to send seg: %d",  ret);
-		k_sem_give(&ch->tx.credits);
+	/* Channel may have been disconnected while sending */
+	if (!ch->chan.conn) {
+		net_buf_unref(seg);
+		return -ECONNRESET;
+	}
```

Comment:
This looks weird. At this point you haven't checked the return value of `bt_l2cap_send_cb()` yet you assume this code still owns the reference?

Btw, this logic seems to go against what we discussed on slack (to be fair though I didn't see you comment on the issue there): we need to establish a consistent behaviour across all APIs when it comes to buffer ownership and errors. My proposal was that APIs taking buffers would always assume ownership of them, so if the calling code doesn't care about errors (e.g. can't do anything useful with them) it doesn't have to check them. This would mean that if the calling code wants to retain a reference (e.g. for retransmission attempts) it needs to do net_buf_ref first.

Reply:
Right it should have unref it on bt_conn_send_cb since that is checking the state of the connection, though when sending the original buffer an extra reference is done by create_seg, I can fix that so that we always make a reference to the original buffer before creating a segment.